compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -7365,18 +7365,25 @@ private void reorderArguments(BLangInvocation iExpr) {
                 return;
             }
 
+            BArrayType restParamArrayType = (BArrayType) invokableSymbol.restParam.type;
+            BType elemType = restParamArrayType.eType;
+
             BLangArrayLiteral arrayLiteral = (BLangArrayLiteral) TreeBuilder.createArrayLiteralExpressionNode();
+            arrayLiteral.type = restParamArrayType;
             List<BLangExpression> exprs = new ArrayList<>();
-
-            BArrayType arrayType = (BArrayType) invokableSymbol.restParam.type;
-            BType elemType = arrayType.eType;
-
             for (BLangExpression restArg : restArgs) {
-                exprs.add(addConversionExprIfRequired(restArg, elemType));
+                // If the restArg is an array (where expected elemType is not an array),
+                // we have to flatten that and create a single array.
+                if (elemType.tag != TypeTags.ARRAY && restArg.getKind() == NodeKind.ARRAY_LITERAL_EXPR) {
```

Comment:
Shall we check `elemType.tag != TypeTags.ARRAY` outside the for loop?

Reply:
You mean outside of `for (BLangExpression restArg : restArgs) {}`?